pyqmc/method/ensemble_optimization_wfbywf.py

```diff
@@ -223,6 +225,142 @@ def renormalize(wfs, norms, pivot=0, N=1):
+                wf,
+                configs_ensemble[wfi][sub_iteration],
+                accumulators={"": transform.onewf()},
+                client=client,
+                npartitions=npartitions,
+                **vmc_kwargs,
+            )
+            data_weighted_ensemble[wfi][sub_iteration], data_unweighted_ensemble[wfi][sub_iteration], configs_ensemble[wfi][sub_iteration] = pyqmc.method.sample_many.sample_overlap(
+                wfs[0 : wfi + 1],
+                configs_ensemble[wfi][sub_iteration],
+                transform.allwfs(),
+                client=client,
+                npartitions=npartitions,
+                **vmc_kwargs,
+            )
+    return data_sample1_ensemble, data_weighted_ensemble, data_unweighted_ensemble, configs_ensemble
+
+
+def evaluate_gradients_threaded(
+    wfs, 
+    configs_ensemble, 
+    updater,
+    wf_start,
+    sub_iteration_offset,
+    client=None, 
+    npartitions=1, 
+    **vmc_kwargs,
+):
+    """Evaluate parameter gradients for each state with threader
+    It runs Monte Carlo evaluations for all the states asynchronously and gathers the results as they complete
+    Each state's vmc evaluations get parallelized using an even share of npartitions
+
+    :parameter list wfs: list of optimized wave functions
+    :parameter list configs_ensemble: nested list of initial configurations indexed by state then by sub-iteration
+    :parameter list updater: nested list of StochasticReconfigurationWfbyWf accumulators indexed by state then by sub-iteration
+    :parameter int wf_start: index for the lowest state in include in Monte Carlo evaluations
+    :parameter int sub_iteration_offset: sub iteration offset for wf_start
+    :parameter client: an object with submit() functions that return futures
+    :parameter int npartitions: the number of workers to submit at a time
+    :parameter dict vmc_kwargs: a dictionary of options for the vmc method
+
+    :return dict data_sample1_ensemble: nested list of vmc outputs indexed by state then by sub-iteration
+    :return dict data_weighted_ensemble: nested list of weighted sample_overlap outputs indexed by state then by sub-iteration
+    :return dict data_unweighted_ensemble: nested list of unweighted sample_overlap outputs indexed by state then by sub-iteration
+    :return list configs_ensemble: nested list of updated configurations indexed by state then by sub-iteration
+    """
+    nwf = len(wfs)
+    sub_iteration_offsets_ensemble = [0] * nwf
+    sub_iteration_offsets_ensemble[wf_start] = sub_iteration_offset
+    nthreads = sum([len(updater[wfi]) - sub_iteration_offsets_ensemble[wfi] for wfi in range(wf_start, nwf)])
+    data_sample1_ensemble = [[0 for _ in range(len(updater[wfi]))] for wfi in range(nwf)]
+    data_weighted_ensemble = [[0 for _ in range(len(updater[wfi]))] for wfi in range(nwf)]
+    data_unweighted_ensemble = [[0 for _ in range(len(updater[wfi]))] for wfi in range(nwf)]
+    result_to_thread = {}
```

Comment:
Make energy_workers and wf_workers, and don't block until the very end. So as_completed only gets called after all the for loops.